File: Linked_List/Python/Find the Minimum and Maximum Number of Nodes Between Critical Points.py
# Time: O(n)
# Space: O(1)
# Problem link: https://leetcode.com/problems/find-the-minimum-and-maximum-number-of-nodes-between-critical-points/description/

class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        prev, temp, count, f_point, p_point, min_maxima = head.val, head.next, 1, -1, -1, inf

        while temp.next:
            if (temp.val > prev and temp.val > temp.next.val) or (temp.val < prev and temp.val < temp.next.val):
                f_point = count if f_point == -1 else f_point
                min_maxima = min(min_maxima, count - p_point) if p_point != -1 else min_maxima
                p_point = count
            prev, count, temp = temp.val, count + 1, temp.next
        
        return [-1, -1] if p_point == f_point else [min_maxima, p_point - f_point]

File: Linked_List/Java/Find the Minimum and Maximum Number of Nodes Between Critical Points.java
// Time: O(n)
// Space: O(1)
// Problem link: https://leetcode.com/problems/find-the-minimum-and-maximum-number-of-nodes-between-critical-points/description/

class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        ListNode temp = head;
        int prev = head.val, count = 1, f_point = -1, p_point = -1, min_maxima = Integer.MAX_VALUE;

        while(temp.next != null) {
            if((temp.val > prev && temp.val > temp.next.val) || (temp.val < prev && temp.val < temp.next.val)) {
                f_point = f_point == -1 ? count : f_point;
                min_maxima = (p_point != -1) ? Math.min(min_maxima, count - p_point) : min_maxima;
                p_point = count;
            }
            prev = temp.val;
            count++;
            temp = temp.next;
        }

        if(p_point == f_point){
            return new int[]{-1, -1};
        } 
        return new int[]{min_maxima, p_point - f_point};
    }
}
